oauthlib/oauth2/rfc6749/parameters.py

```diff
@@ -118,10 +137,21 @@ def prepare_token_request(grant_type, body='', **kwargs):
     if 'scope' in kwargs:
         kwargs['scope'] = list_to_scope(kwargs['scope'])
 
+    # pull the `client_id` out of the kwargs. 
+    client_id = kwargs.pop('client_id', None)
+    if include_client_id:
+        if client_id is not None:
+            params.append((unicode_type('client_id'), client_id))
+
     for k in kwargs:
+        # this handles: `code`, `redirect_uri`, or undocumented params
         if kwargs[k]:
             params.append((unicode_type(k), kwargs[k]))
 
+    if ('client_secret' in kwargs) and ('client_secret' not in params):
```

Comment:
This might deserve a comment...

Reply:
good idea.

Reply:
wow, thanks for this comment. i actually found a weird issue from this and am rewriting the function a bit!

Reply:
fixed, and with a comment.

the outdated code will *never* find "client_secret" in params, because params is a list of tuples. 

    ('client_secret' not in params)

the search should have been something like this:

    ('client_secret' not in [p[0] for p in params])

but that's ugly and inefficient.

in the fix, I just popped client_secret from the kwargs and compared it to `None`.